File: android/crosspl/anno/src/main/kotlin/org/elastos/tools/crosspl/processor/generator/CrossProxyGenerator.kt
package org.elastos.tools.crosspl.processor.generator

import org.elastos.tools.crosspl.processor.*
import java.io.File

class CrossProxyGenerator {
    companion object {
        fun Generate(crossProxyDir: File, classInfo: CrossClassInfo): Boolean {
            val proxyHeaderFile = GetHeaderFile(crossProxyDir, classInfo)
            val proxySourceFile = GetSourceFile(crossProxyDir, classInfo)

            var ret = GenerateHeader(proxyHeaderFile, classInfo)
            if(! ret) {
                return ret
            }

            ret = GenerateSource(proxySourceFile, classInfo)
            if(! ret) {
                return ret
            }

            return true
        }

        fun GetSourceFile(crossProxyDir: File, classInfo: CrossClassInfo): File {
            return File(crossProxyDir, classInfo.cppInfo.className + ".proxy.cpp")
        }

        fun GetHeaderFile(crossProxyDir: File, classInfo: CrossClassInfo): File {
            return File(crossProxyDir, classInfo.cppInfo.className + ".proxy.hpp")
        }

        private fun GenerateHeader(proxyFile: File, classInfo: CrossClassInfo): Boolean {
            Log.w("Generate: ${proxyFile.absolutePath}")

            val tmpl = CrossTmplUtils.ReadTmplContent(CrossClassProxyHeaderTmpl)
            val content = classInfo.makeProxyDeclare(tmpl)
            CrossTmplUtils.WriteContent(proxyFile, content)

            return true
        }

        private fun GenerateSource(proxyFile: File, classInfo: CrossClassInfo): Boolean {
            Log.w("Generate: ${proxyFile.absolutePath}")

            val tmpl = CrossTmplUtils.ReadTmplContent(CrossClassProxySourceTmpl)
            val content = classInfo.makeProxySource(tmpl)
            CrossTmplUtils.WriteContent(proxyFile, content)

            return true
        }


        private const val CrossClassProxyHeaderTmpl = "/CrossPLClass.proxy.hpp.tmpl"
        private const val CrossClassProxySourceTmpl = "/CrossPLClass.proxy.cpp.tmpl"
    }
}

File: ios/crosspl/anno/swift/generator/CrossProxyGenerator.swift
import Foundation

open class CrossProxyGenerator {
  static func Generate(crossProxyDir: URL, classInfo: CrossClassInfo) -> Bool {
    let proxyHeaderFile = GetHeaderFile(crossProxyDir: crossProxyDir, classInfo: classInfo)
    let proxySourceFile = GetSourceFile(crossProxyDir: crossProxyDir, classInfo: classInfo)
    
    var ret = GenerateHeader(proxyFile: proxyHeaderFile, classInfo: classInfo)
    if ret == false {
      return ret
    }
    
    ret = GenerateSource(proxyFile: proxySourceFile, classInfo: classInfo)
    if ret == false {
      return ret
    }

    
    return true
  }
  
  static func GetSourceFile(crossProxyDir: URL, classInfo: CrossClassInfo) -> URL {
    return crossProxyDir.appendingPathComponent(classInfo.cppInfo.className! + ".proxy.mm")
  }
  
  static func GetHeaderFile(crossProxyDir: URL, classInfo: CrossClassInfo) -> URL {
    return crossProxyDir.appendingPathComponent(classInfo.cppInfo.className! + ".proxy.h")
  }

  private static func GenerateHeader(proxyFile: URL, classInfo: CrossClassInfo) -> Bool {
    let sourceChanged = checkSourceChanged(proxyFile: proxyFile, classInfo: classInfo)
    if sourceChanged == false {
      print("Ignore Generate: \(proxyFile.path)")
      return true
    }
    
    print("Generate: \(proxyFile.path)")
    let tmpl = CrossTmplUtils.ReadTmplContent(tmplName: CrossClassProxyHeaderTmpl)
    let content = classInfo.makeProxyDeclare(tmpl: tmpl)
    CrossTmplUtils.WriteContent(file: proxyFile, content: content)
  
    return true
  }

  private static func GenerateSource(proxyFile: URL, classInfo: CrossClassInfo) -> Bool {
    let sourceChanged = checkSourceChanged(proxyFile: proxyFile, classInfo: classInfo)
    if sourceChanged == false {
      print("Ignore Generate: \(proxyFile.path)")
      return true
    }
    
    print("Generate: \(proxyFile.path)")
    let tmpl = CrossTmplUtils.ReadTmplContent(tmplName: CrossClassProxySourceTmpl)
    let content = classInfo.makeProxySource(tmpl: tmpl)
    CrossTmplUtils.WriteContent(file: proxyFile, content: content)
  
    return true
  }

  private static func checkSourceChanged(proxyFile: URL, classInfo: CrossClassInfo) -> Bool {
    let srcAttribs = try? FileManager.default.attributesOfItem(atPath: classInfo.sourcePath!)
    let srcModifyTime = srcAttribs?[FileAttributeKey.modificationDate] as? Date
    let proxyAttribs = try? FileManager.default.attributesOfItem(atPath: proxyFile.path)
    let proxyModifyTime = proxyAttribs?[FileAttributeKey.modificationDate] as? Date
    if srcModifyTime != nil
    && proxyModifyTime != nil
    && srcModifyTime! < proxyModifyTime! { // source file not changed
      return false
    }
    
    return true
  }
  
  private static let CrossClassProxyHeaderTmpl = "/CrossPLClass.proxy.h.tmpl"
  private static let CrossClassProxySourceTmpl = "/CrossPLClass.proxy.mm.tmpl"
}
